File: app/src/main/java/app/fit/fitndflow/ui/features/home/HomeFragment.java
package app.fit.fitndflow.ui.features.home;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static app.fit.fitndflow.ui.features.common.notification.MyNotificationManager.scheduleNotification;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.viewbinding.ViewBinding;

import com.fit.fitndflow.R;
import com.fit.fitndflow.databinding.AddSerieTrainingFragmentBinding;
import com.fit.fitndflow.databinding.DialogCreationInputBinding;
import com.fit.fitndflow.databinding.MainListFragmentBinding;

import java.util.Date;
import java.util.List;

import app.fit.fitndflow.data.common.SharedPrefs;
import app.fit.fitndflow.domain.Utils;
import app.fit.fitndflow.domain.model.CategoryModel;
import app.fit.fitndflow.domain.model.ExerciseModel;
import app.fit.fitndflow.ui.features.categories.CategoriesListFragment;
import app.fit.fitndflow.ui.features.common.CommonFragment;
import app.fit.fitndflow.ui.features.common.notification.MyNotificationManager;
import app.fit.fitndflow.ui.features.training.AddSerieTrainingFragment;

public class HomeFragment extends CommonFragment<HomeViewModel> implements ExerciseClickCallback {
    private MainListFragmentBinding binding;

    @Override
    protected Class getViewModelClass() {
        return HomeViewModel.class;
    }

    private HomeViewModel homeViewModel;

    private boolean isShownNotificationConfiguration = false;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = MainListFragmentBinding.inflate(getLayoutInflater());
        View view = binding.getRoot();
        super.onCreateView(inflater, container, savedInstanceState);
        if(!isShownNotificationConfiguration){
            MyNotificationManager.askForPermissions(this);
            isShownNotificationConfiguration = true;
        }
        scheduleNotification(requireActivity(), 6000, MyNotificationManager.TRAINING_TYPE);

        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (SharedPrefs.getApikeyFromSharedPRefs(requireContext()) == null) {
            viewModel.requestRegisterEmptyUser(requireContext());
        }
        setViewModelObservers();
        setClickListeners();
    }

    private void  printExercises(List<CategoryModel> categoryModelList){
        binding.emptyTxt.setVisibility(GONE);
        binding.parentContainer.removeAllViews();
        for (CategoryModel category : categoryModelList) {
            CategoryCustomView  categoryView = new CategoryCustomView(getContext(), category, this);
            binding.parentContainer.addView(categoryView);
        }
    }

    private void printEmptyView(){
        binding.emptyTxt.setVisibility(VISIBLE);
        binding.parentContainer.removeAllViews();
    }

    private void setClickListeners() {
        binding.btnLeft.setOnClickListener(view ->{
            if(!homeViewModel.getIsLoading().getValue()){
                viewModel.dayBefore();
            }
        });
        binding.btnRight.setOnClickListener(view ->{
            if(!homeViewModel.getIsLoading().getValue()) {
                viewModel.dayAfter();
            }
        });
        binding.buttonPanel.setOnClickListener(view ->{
            if(!homeViewModel.getIsLoading().getValue()) {
                addFragment(new CategoriesListFragment());
            }
        });
    }

    private void setViewModelObservers() {
        homeViewModel = ViewModelProviders.of(getActivity()).get(HomeViewModel.class);

        final Observer<List<CategoryModel>> observer = new Observer<List<CategoryModel>>() {
            @Override
            public void onChanged(List<CategoryModel> categories) {
                if(categories == null || categories.isEmpty()){
                    binding.buttonPanel.setText("Añadir entrenamiento");
                    printEmptyView();
                } else {
                    binding.buttonPanel.setText("Añadir ejercicio");
                    printExercises(categories);
                }
            }
        };
        homeViewModel.getDailyTrainingMutableList().observe(getActivity(), observer);

        final Observer<Boolean> observerLoading = new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean isLoading) {
                try{
                    if (isLoading) {
                        showLoading();

                    } else {
                        hideLoading();
                    }
                }catch(Exception exception){
                    Log.e("Error","show loading");
                }
            }
        };
        homeViewModel.getIsLoading().observe(getActivity(), observerLoading);


        final Observer<Date> actualDateObserver = new Observer<Date>() {
            @Override
            public void onChanged(Date date) {
                homeViewModel.requestTrainingFromModel(requireContext());

                if (Utils.isYesterday(date)) {
                    binding.dateName.setText(R.string.yesterday_date_selector);
                    binding.dayOfWeek.setVisibility(GONE);
                } else if (Utils.isToday(date)) {
                    binding.dateName.setText(R.string.today_date_selector);
                    binding.dayOfWeek.setVisibility(GONE);
                } else if (Utils.isTomorrow(date)) {
                    binding.dateName.setText(R.string.tomorrow_date_selector);
                    binding.dayOfWeek.setVisibility(GONE);
                } else {
                    binding.dateName.setText(Utils.getCalendarFormatDate(date));
                    binding.dayOfWeek.setText(Utils.dayOfWeek(date));
                    binding.dayOfWeek.setVisibility(VISIBLE);
                }
            }
        };
        viewModel.getMutableActualDate().observe(getActivity(), actualDateObserver);
        //observing error
        final Observer<Boolean> errorObserver = new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean isError) {
                if (isError) {
                    showBlockError();
                    viewModel.getMutableFullScreenError().setValue(false);
                }
            }
        };
        viewModel.getMutableFullScreenError().observe(getActivity(), errorObserver);
    }

    @Override
    public void showExerciseTrainingDetail(ExerciseModel exercise) {
        addFragment(AddSerieTrainingFragment.newInstance(exercise));
    }
}

File: app/src/main/java/app/fit/fitndflow/ui/features/home/HomeFragment.kt
package app.fit.fitndflow.ui.features.home

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.lifecycleScope
import app.fit.fitndflow.data.common.SharedPrefs
import app.fit.fitndflow.domain.Utils
import app.fit.fitndflow.domain.model.CategoryModel
import app.fit.fitndflow.domain.model.ExerciseModel
import app.fit.fitndflow.ui.features.categories.CategoriesListFragment
import app.fit.fitndflow.ui.features.common.CommonFragment
import app.fit.fitndflow.ui.features.common.notification.MyNotificationManager
import app.fit.fitndflow.ui.features.common.notification.MyNotificationManager.scheduleNotification
import app.fit.fitndflow.ui.features.training.AddSerieTrainingFragment
import com.fit.fitndflow.R
import com.fit.fitndflow.databinding.MainListFragmentBinding
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import java.util.Calendar
import java.util.Calendar.DAY_OF_YEAR
import java.util.Calendar.HOUR_OF_DAY
@AndroidEntryPoint
class HomeFragment : CommonFragment(), ExerciseClickCallback {

    private lateinit var binding: MainListFragmentBinding
    private val homeViewModel: HomeViewModel by activityViewModels()
    private var isShownNotificationConfiguration: Boolean = false
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = MainListFragmentBinding.inflate(layoutInflater)
        val myView = binding.root
        super.onCreateView(inflater, container, savedInstanceState)
        if (!isShownNotificationConfiguration) {
            MyNotificationManager.askForPermissions(this)
            isShownNotificationConfiguration = true
        }
        scheduleNotification()

        return myView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        attachObservers()
        setClickListeners()
    }

    private fun requestRegisterOrRequestTraining(){
        if (SharedPrefs.getApikeyFromSharedPRefs(requireContext()) == null) {
            homeViewModel.requestRegisterEmptyUser()
        } else {
            homeViewModel.requestTrainingFromModel()
        }
    }

    override fun onResume() {
        homeViewModel.emitDate()
        super.onResume()
    }

    private fun scheduleNotification() {
        val calendar: Calendar = Calendar.getInstance()
        calendar.add(DAY_OF_YEAR, 1)
        calendar.add(HOUR_OF_DAY, -1)
        val tomorrowInMillis: Long = calendar.timeInMillis
        scheduleNotification(
            requireActivity(),
            tomorrowInMillis,
            MyNotificationManager.TRAINING_TYPE
        )
    }

    private fun printExercises(categoryModelList: List<CategoryModel>) {
        binding.apply {
            emptyTxt.visibility = GONE
            parentContainer.removeAllViews()
            for (category: CategoryModel in categoryModelList) {
                val categoryView = CategoryCustomView(context, category, this@HomeFragment)
                parentContainer.addView(categoryView)
            }
        }
    }

    private fun printEmptyView() {
        binding.apply {
            emptyTxt.visibility = VISIBLE
            parentContainer.removeAllViews()
        }
    }

    private fun setClickListeners() {
        binding.apply {
            btnLeft.setOnClickListener {
                homeViewModel.dayBefore()
            }
            btnRight.setOnClickListener {
                homeViewModel.dayAfter()
            }
            buttonPanel.setOnClickListener {
                addFragment(CategoriesListFragment())
            }
        }
    }

    private fun attachObservers() {
        homeViewModel.state.onEach(::handleState).launchIn(viewLifecycleOwner.lifecycleScope)
    }

    private fun handleState(state: State) {
        when (state) {
            is State.CurrentDateChanged -> {
                val date = state.date
                requestRegisterOrRequestTraining()
                binding.apply {
                    if (Utils.isYesterday(date)) {
                        dateName.setText(R.string.yesterday_date_selector)
                        dayOfWeek.visibility = GONE
                    } else if (Utils.isToday(date)) {
                        dateName.setText(R.string.today_date_selector)
                        dayOfWeek.visibility = GONE
                    } else if (Utils.isTomorrow(date)) {
                        dateName.setText(R.string.tomorrow_date_selector)
                        dayOfWeek.visibility = GONE
                    } else {
                        dateName.setText(Utils.getCalendarFormatDate(date))
                        dayOfWeek.setText(Utils.dayOfWeek(date))
                        dayOfWeek.visibility = VISIBLE
                    }
                }
            }

            State.RegisterCompleted -> {
                hideLoading()
            }

            State.FullScreenError -> {
                hideLoading()
                showBlockError()
            }

            State.Loading -> {
                try {
                    showLoading()
                } catch (exception: Exception) {
                    Log.e("Error", "show loading")
                }
            }

            is State.TrainingListRecived -> {
                hideLoading()
                val categories = state.categoryList
                if (categories == null || categories.isEmpty()) {
                    binding.buttonPanel.setText(R.string.add_training)
                    printEmptyView()
                } else {
                    binding.buttonPanel.setText(R.string.add_exercise)
                    printExercises(categories)
                }
            }
        }
    }

    override fun showExerciseTrainingDetail(exercise: ExerciseModel) {
        addFragment(AddSerieTrainingFragment.newInstance(exercise))
    }
}